### ISSUE TITLE
Render to OpenGL Target

### ISSUE THREAD

**Issue by J-Cake** [samizdatco/skia-canvas#8]
I've been trying to do this for a while now - I've got an OpenGL Context with glfwJS created, and I'd like to use Skia to draw to it, I'm just wondering if there's a way to make the two interoperable? I realise there's a lot to bridge, but according to [this post](https://groups.google.com/forum/#!topic/skia-discuss/P4GO92rxIaM), it seems doable.

**Comment by J-Cake** [comment]
Further digging has revealed it's fairly doable, although I'm not sure how it would be integrated with NodeJS:

https://gist.github.com/ad8e/dd150b775ae6aa4d5cf1a092e4713add

**Comment by konsumer** [comment]
I am also very interested in this. Has anyone figured it out?

**Comment by lucasmerlin** [comment]
Skia seems to have some slight differences when rendering on the cpu vs the gpu and I need to perfectly match how a browser would render the image. So when I'm rendering with skia-canvas, it would be great for skia-canvas to support rendering on the gpu!
I saw that there is a gpu-render branch and I was able to modify it to use vulkan to be able to run it in a docker container:
https://github.com/lucasmerlin/skia-canvas/tree/gpu-render
Using this in production right now and it's working great so far!
I can also share some more details of how I was able to get this to run in docker, if anyone is curious about that. 

Are there any plans to actually release a version with gpu support to npm?

**Comment by samizdatco** [comment]
I've been able to get things working well on the Mac (both for drawing to windows and running offscreen with the GPU renderer). The reason I've been wary of trying to integrate the changes is two-fold:
  1. It's unclear to me whether it will be possible to ship a single version of the library that switches between CPU and GPU modes in a reliable way (and without bloating either the size or the system requirements)
  2. I'm totally out-of-touch with how OpenGL works on Windows and Linux at this point. 

It's encouraging to hear you've been able to get the GPU renderer working in Docker containers, @lucasmerlin! In your fork it looks like you're using the Vulkan backend. I'd been thinking of sticking with OpenGL since it stood the best chance of running on all three supported platforms (and using something like [glutin](https://crates.io/crates/glutin) for windowing). Can you explain a bit about how Gl/Vulkan/Mesa and X11/Wayland interact? It's unclear to me what libraries need to be accounted for.

I'd also be curious to see some Dockerfiles with the necessary dependencies for a) offscreen GPU rendering, and b) opening a window via glutin or some other cross-platform abstraction.

And then there's Windows. It'd be great if someone could test out whether using glutin for windowing and GL for rendering is a reasonable option there as well.

**Comment by konsumer** [comment]
Let me know what I can do to help. I am strong on the js-side, weak on the rs-side, and am happy to help with anything I can.

**Comment by samizdatco** [comment]
Building off of @lucasmerlin's fine work, I've gotten [GPU-based rendering](https://github.com/samizdatco/skia-canvas/tree/gpu) working on macOS (using Metal) and Linux (using Vulkan). My understanding is that the Vulkan backend should work on Windows as well, but the VMs I've tested it on don't have GPUs. So I could use some help verifying whether it's actually functioning properly from someone with an actual Windows box to test on. I'm also curious to know what additional dependencies are needed to get this version running in common Docker containers.

There's a new `.gpu` attribute on the **Canvas** object which will be `true` if it was able to set up a hardware-based renderer (and can be set to `false` to switch back to the CPU renderer manually). Before merging the pull request I'd love it if people could kick the tires and let me know if they find any regressions compared to the software renderer. As far as I can tell though, the changes are nearly all positive and go a long way toward addressing #88, #85, and #38.

You can install the current release candidate (rc2) via:
```console
npm install https://github.com/samizdatco/skia-canvas#gpu
```

**Comment by samizdatco** [comment]
I've made some progress on the rendering-to-a-window front, but haven't totally solved the problem and could greatly benefit from some suggestions at this point.

Here's a new repository of [Windowing Prototypes](https://github.com/samizdatco/skia-window-tests) where I've been attempting to go from Skia to the screen on all the platforms we need to support. You can run the tests in each of the subfolders by cd'ing into the `metal`, `gl`, or `vulkan` directory and typing `cargo run`.

So far, the only one that's fully functional is the Mac version using the Metal backend. I'd like to find a single solution that works for Linux & Windows which effectively means either OpenGL or Vulkan, but have only gotten either of them part of the way there.

I could really use some help debugging the two dead-ends I've reached:
  1. Figuring out how to keep the ‘current’ GL context tied to the correct window (and its Skia context)
  2. Fixing the validation errors in the Vulkan render pass that are preventing the graphics pipeline from completing

Both of these issues are fleshed out in more detail in the repo's Readme, along with pointers to the spots in the code where I think the problems (and solutions?) may lie...

**Comment by konsumer** [comment]
I can verify Mac/Metal demo runs great.
<img width="855" alt="Screen Shot 2022-07-23 at 3 07 24 PM" src="https://user-images.githubusercontent.com/83857/180624431-25ab964b-fb25-4d43-b86e-86cee8547293.png">

OpenGL runs great on my mac:
<img width="841" alt="Screen Shot 2022-07-23 at 3 09 05 PM" src="https://user-images.githubusercontent.com/83857/180624487-702d70a4-48bb-42d6-8417-211424c5a37f.png">

But like you said, window-resize or close makes the other windows weird. What if we just didn't allow window resize after init? What if we only did a single window?

If I change code to show just 1 window, it works great, resize is fine, and it seems ok to close it.

Vulcan crashes (which is not at all surprising on Mac), so maybe focusing on OpenGL makes sense.

**Comment by samizdatco** [comment]
One thing I forgot to mention about Vulkan: to use it on the Mac you'll first need to [install the Vulkan SDK](https://www.lunarg.com/vulkan-sdk/) and set some environment variables before running the test. It'll look something like:
```sh
export VULKAN_SDK="/path/to/vulkan-sdk/macOS"
export DYLD_LIBRARY_PATH="$VULKAN_SDK/lib:$DYLD_LIBRARY_PATH"
export VK_ICD_FILENAMES="$VULKAN_SDK/etc/vulkan/icd.d/MoltenVK_icd.json"
export VK_LAYER_PATH="$VULKAN_SDK/etc/vulkan/explicit_layer.d"
```

I also found that I needed to create the `etc` directory myself and make a symlink to `share/vulkan` to satisfy the compiler:
```console
❯ ls -l macOS/etc/vulkan
lrwxr-xr-x 1 cds admin 16 Jul 19 22:11 macOS/etc/vulkan -> ../share/vulkan/
```

**Comment by konsumer** [comment]
Ok, got vulcan SDK installed, and setup right, and I did the symlink to etc. I get a black window and lots of output:

```
    Finished dev [unoptimized + debuginfo] target(s) in 0.17s
     Running `target/debug/skpn-win`
[2022-07-23T23:17:40Z INFO  skulpin_renderer::instance] Finding vulkan entry point
[2022-07-23T23:17:40Z INFO  skulpin_renderer::instance] Found Vulkan version: (1, 3, 216)
[2022-07-23T23:17:40Z DEBUG skulpin_renderer::instance] Available Layers: [
        LayerProperties {
            layer_name: "VK_LAYER_LUNARG_api_dump",
            spec_version: 4206808,
            implementation_version: 2,
            description: "LunarG API dump layer",
        },
        LayerProperties {
            layer_name: "VK_LAYER_KHRONOS_profiles",
            spec_version: 4206808,
            implementation_version: 1,
            description: "Khronos Profiles layer",
        },
        LayerProperties {
            layer_name: "VK_LAYER_KHRONOS_validation",
            spec_version: 4206808,
            implementation_version: 1,
            description: "Khronos Validation Layer",
        },
        LayerProperties {
            layer_name: "VK_LAYER_KHRONOS_synchronization2",
            spec_version: 4206808,
            implementation_version: 1,
            description: "Khronos Synchronization2 layer",
        },
    ]
[2022-07-23T23:17:40Z DEBUG skulpin_renderer::instance] Available Extensions: [
        ExtensionProperties {
            extension_name: "VK_KHR_device_group_creation",
            spec_version: 1,
        },
        ExtensionProperties {
            extension_name: "VK_KHR_external_fence_capabilities",
            spec_version: 1,
        },
        ExtensionProperties {
            extension_name: "VK_KHR_external_memory_capabilities",
            spec_version: 1,
        },
        ExtensionProperties {
            extension_name: "VK_KHR_external_semaphore_capabilities",
            spec_version: 1,
        },
        ExtensionProperties {
            extension_name: "VK_KHR_get_physical_device_properties2",
            spec_version: 2,
        },
        ExtensionProperties {
            extension_name: "VK_KHR_get_surface_capabilities2",
            spec_version: 1,
        },
        ExtensionProperties {
            extension_name: "VK_KHR_surface",
            spec_version: 25,
        },
        ExtensionProperties {
            extension_name: "VK_EXT_debug_report",
            spec_version: 10,
        },
        ExtensionProperties {
            extension_name: "VK_EXT_debug_utils",
            spec_version: 2,
        },
        ExtensionProperties {
            extension_name: "VK_EXT_metal_surface",
            spec_version: 1,
        },
        ExtensionProperties {
            extension_name: "VK_EXT_swapchain_colorspace",
            spec_version: 4,
        },
        ExtensionProperties {
            extension_name: "VK_MVK_macos_surface",
            spec_version: 3,
        },
        ExtensionProperties {
            extension_name: "VK_KHR_portability_enumeration",
            spec_version: 1,
        },
    ]
[2022-07-23T23:17:40Z DEBUG skulpin_renderer::instance] Using layers: ["VK_LAYER_KHRONOS_validation"]
[2022-07-23T23:17:40Z DEBUG skulpin_renderer::instance] Using extensions: ["VK_KHR_surface", "VK_EXT_metal_surface", "VK_KHR_portability_enumeration", "VK_KHR_get_physical_device_properties2", "VK_EXT_swapchain_colorspace", "VK_EXT_debug_utils"]
[2022-07-23T23:17:40Z INFO  skulpin_renderer::instance] Creating vulkan instance
[2022-07-23T23:17:40Z INFO  skulpin_renderer::instance] Seting up vulkan debug callback
[2022-07-23T23:17:40Z INFO  skulpin_renderer::device] Preferred device types: [DiscreteGpu, IntegratedGpu]
[2022-07-23T23:17:40Z INFO  skulpin_renderer::device] Available queue families:
[2022-07-23T23:17:40Z INFO  skulpin_renderer::device] Queue Family 0
[2022-07-23T23:17:40Z INFO  skulpin_renderer::device] QueueFamilyProperties {
        queue_flags: GRAPHICS | COMPUTE | TRANSFER,
        queue_count: 1,
        timestamp_valid_bits: 64,
        min_image_transfer_granularity: Extent3D {
            width: 1,
            height: 1,
            depth: 1,
        },
    }
[2022-07-23T23:17:40Z INFO  skulpin_renderer::device] Queue Family 1
[2022-07-23T23:17:40Z INFO  skulpin_renderer::device] QueueFamilyProperties {
        queue_flags: GRAPHICS | COMPUTE | TRANSFER,
        queue_count: 1,
        timestamp_valid_bits: 64,
        min_image_transfer_granularity: Extent3D {
            width: 1,
            height: 1,
            depth: 1,
        },
    }
[2022-07-23T23:17:40Z INFO  skulpin_renderer::device] Queue Family 2
[2022-07-23T23:17:40Z INFO  skulpin_renderer::device] QueueFamilyProperties {
        queue_flags: GRAPHICS | COMPUTE | TRANSFER,
        queue_count: 1,
        timestamp_valid_bits: 64,
        min_image_transfer_granularity: Extent3D {
            width: 1,
            height: 1,
            depth: 1,
        },
    }
[2022-07-23T23:17:40Z INFO  skulpin_renderer::device] Queue Family 3
[2022-07-23T23:17:40Z INFO  skulpin_renderer::device] QueueFamilyProperties {
        queue_flags: GRAPHICS | COMPUTE | TRANSFER,
        queue_count: 1,
        timestamp_valid_bits: 64,
        min_image_transfer_granularity: Extent3D {
            width: 1,
            height: 1,
            depth: 1,
        },
    }
[2022-07-23T23:17:40Z INFO  skulpin_renderer::device] Graphics QF: Some(0)  Present QF: Some(0)
[2022-07-23T23:17:40Z INFO  skulpin_renderer::device] Found suitable device 'Apple M1 Pro' API: 1.1.216 DriverVersion: 0.2.1918 Score = 100
[2022-07-23T23:17:40Z DEBUG skulpin_renderer::debug_reporter] "Loading layer library /Users/konsumer/VulkanSDK/1.3.216.0/macOS//etc/vulkan/explicit_layer.d/../../../lib/libVkLayer_khronos_validation.dylib"
[2022-07-23T23:17:40Z INFO  skulpin_renderer::debug_reporter] "Inserted device layer VK_LAYER_KHRONOS_validation (/Users/konsumer/VulkanSDK/1.3.216.0/macOS//etc/vulkan/explicit_layer.d/../../../lib/libVkLayer_khronos_validation.dylib)"
[2022-07-23T23:17:40Z INFO  skulpin_renderer::debug_reporter] "       Using \"Apple M1 Pro\" with driver: \"/Users/konsumer/VulkanSDK/1.3.216.0/macOS//etc/vulkan/icd.d/../../../lib/libMoltenVK.dylib\"\n"
[2022-07-23T23:17:40Z INFO  skulpin_renderer::debug_reporter] "Using MTLEvent for Vulkan semaphores."
[2022-07-23T23:17:41Z INFO  skulpin_renderer::debug_reporter] "Created VkDevice to run on GPU Apple M1 Pro with the following 2 Vulkan extensions enabled:\n\t\tVK_KHR_portability_subset v1\n\t\tVK_KHR_swapchain v70"
[2022-07-23T23:17:41Z INFO  skulpin_renderer::debug_reporter] "Validation Information: [ UNASSIGNED-cache-file-error ] Object 0: handle = 0x13b8e0a18, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xf0bb3995 | Cannot open shader validation cache at /Users/konsumer/.cache/shader_validation_cache.bin for reading (it may not exist yet)"
[2022-07-23T23:17:41Z INFO  skulpin_renderer::skia_support] Setting up skia backend context with queue family index 0
[2022-07-23T23:17:41Z INFO  skulpin_renderer::swapchain] Surface format: SurfaceFormatKHR { format: B8G8R8A8_UNORM, color_space: SRGB_NONLINEAR }
[2022-07-23T23:17:41Z DEBUG skulpin_renderer::swapchain] Swapchain extents chosen by surface capabilities (800 600)
[2022-07-23T23:17:41Z INFO  skulpin_renderer::swapchain] Extents: Extent2D { width: 800, height: 600 }
[2022-07-23T23:17:41Z INFO  skulpin_renderer::swapchain] Available present modes: [FIFO, IMMEDIATE]
[2022-07-23T23:17:41Z INFO  skulpin_renderer::swapchain] Preferred present modes: [Fifo]
[2022-07-23T23:17:41Z INFO  skulpin_renderer::swapchain] Present mode: FIFO
[2022-07-23T23:17:41Z INFO  skulpin_renderer::debug_reporter] "Created 3 swapchain images with initial size (800, 600)."
[2022-07-23T23:17:41Z INFO  skulpin_renderer::skia_renderpass] Creating command pool with queue family index 0
[2022-07-23T23:17:41Z INFO  skulpin_renderer::skia_renderpass] Create skia surfaces with extent: Extent2D { width: 800, height: 600 }
[2022-07-23T23:17:41Z ERROR skulpin_renderer::debug_reporter] "Validation Error: [ UNASSIGNED-CoreValidation-DrawState-InvalidImageLayout ] Object 0: handle = 0x13b72cfd8, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0x4dae5635 | vkQueueSubmit(): pSubmits[0].pCommandBuffers[0] command buffer VkCommandBuffer 0x13b72cfd8[] expects VkImage 0x3a6cbb0000000025[] (subresource: aspectMask 0x1 array layer 0, mip level 0) to be in layout VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL--instead, current layout is VK_IMAGE_LAYOUT_UNDEFINED."
[2022-07-23T23:17:41Z ERROR skulpin_renderer::debug_reporter] "Validation Error: [ UNASSIGNED-CoreValidation-DrawState-InvalidImageLayout ] Object 0: handle = 0x13b72d118, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0x4dae5635 | vkQueueSubmit(): pSubmits[0].pCommandBuffers[0] command buffer VkCommandBuffer 0x13b72d118[] expects VkImage 0xa43473000000002d[] (subresource: aspectMask 0x1 array layer 0, mip level 0) to be in layout VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL--instead, current layout is VK_IMAGE_LAYOUT_UNDEFINED."
[2022-07-23T23:17:41Z ERROR skulpin_renderer::debug_reporter] "Validation Error: [ UNASSIGNED-CoreValidation-DrawState-InvalidImageLayout ] Object 0: handle = 0x13b72d258, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0x4dae5635 | vkQueueSubmit(): pSubmits[0].pCommandBuffers[0] command buffer VkCommandBuffer 0x13b72d258[] expects VkImage 0xa808d50000000033[] (subresource: aspectMask 0x1 array layer 0, mip level 0) to be in layout VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL--instead, current layout is VK_IMAGE_LAYOUT_UNDEFINED."
[2022-07-23T23:17:46Z DEBUG skulpin_renderer::renderer] destroying Renderer
[2022-07-23T23:17:46Z DEBUG skulpin_renderer::skia_renderpass] destroying VkSkiaRenderPass
[2022-07-23T23:17:46Z DEBUG skulpin_renderer::skia_renderpass] destroyed VkSkiaRenderPass
[2022-07-23T23:17:46Z DEBUG skulpin_renderer::swapchain] destroying VkSwapchain
[2022-07-23T23:17:46Z DEBUG skulpin_renderer::swapchain] destroyed VkSwapchain
[2022-07-23T23:17:46Z DEBUG skulpin_renderer::device] destroying VkDevice
[2022-07-23T23:17:46Z DEBUG skulpin_renderer::debug_reporter] "Unloading layer library /Users/konsumer/VulkanSDK/1.3.216.0/macOS//etc/vulkan/explicit_layer.d/../../../lib/libVkLayer_khronos_validation.dylib"
[2022-07-23T23:17:46Z DEBUG skulpin_renderer::device] destroyed VkDevice
[2022-07-23T23:17:46Z DEBUG skulpin_renderer::instance] destroying VkInstance
[2022-07-23T23:17:46Z DEBUG skulpin_renderer::instance] destroyed VkInstance
[2022-07-23T23:17:46Z DEBUG skulpin_renderer::renderer] destroyed Renderer
```

**Comment by samizdatco** [comment]
That looks pretty much identical to what I'm seeing on macOS as well. I'd be curious to see what it does on a Windows or Linux system with a GPU though if anyone can test that and post the output.

**Comment by samizdatco** [comment]
The Vulkan demo in the [Windowing Prototypes repo](https://github.com/samizdatco/skia-window-tests) now works for me on macOS (downgrading MoltenVK to a 1.2.x version did the trick) and I've gotten it to use a similar threading setup to the Metal demo. It'd help a ton if someone could try running the Vulkan demo on Windows and/or Linux and let me know if it does anything unexpected on either of those platforms.

**Comment by lucasmerlin** [comment]
I was finally able to test the gpu branch on windows and all seems good, the tests ran through. Also works with my docker / kubernetes setup, we're using it now on hellopaint.io!

Also tried the windowing repo and it's flickering a lot on windows, but is smooth as long as a resize, when I let go it starts flickering again. All the windows are flickering, even before I resize one window.
https://user-images.githubusercontent.com/8009393/181111898-a2432c5c-52c7-4d73-b92f-87482326d782.mp4

**Comment by samizdatco** [comment]
I've reached the testing phase for [this PR](https://github.com/samizdatco/skia-canvas/pull/108) and would love to know how it works on other systems. The big change is the addition of a `Window` class which encapsulates a `Canvas` and provides event-handling hooks. Documentation for [Window](https://github.com/samizdatco/skia-canvas/tree/gpu-window#window) and the associated [App](https://github.com/samizdatco/skia-canvas/tree/gpu-window#app) global can be found in the branch's readme and other details are spelled out in the PR comment.

You can install the current release candidate (rc5) via:
```console
npm install https://github.com/samizdatco/skia-canvas#gpu-window
```

**Comment by samizdatco** [comment]
The PR has landed in the 1.0 release. Closing this issue but please open new ones with any windowing-specific bugs that crop up.
